scalding-core/src/main/scala/com/twitter/scalding/macros/impl/FieldsProviderImpl.scala

```diff
@@ -80,53 +80,81 @@ object FieldsProviderImpl {
+        }
+        fieldType match {
+          case tpe if tpe =:= typeOf[String] => simpleRet
+          case tpe if tpe =:= typeOf[Boolean] => simpleRet
+          case tpe if tpe =:= typeOf[Short] => simpleRet
+          case tpe if tpe =:= typeOf[Int] => simpleRet
+          case tpe if tpe =:= typeOf[Long] => simpleRet
+          case tpe if tpe =:= typeOf[Float] => simpleRet
+          case tpe if tpe =:= typeOf[Double] => simpleRet
+          case tpe if tpe.erasure =:= typeOf[Option[Any]] && isOption == true => c.abort(c.enclosingPosition, s"Case class ${T} has nested options, not supported currently.")
+          case tpe if tpe.erasure =:= typeOf[Option[Any]] =>
+            val innerType = tpe.asInstanceOf[TypeRefApi].args.head
+            matchField(innerType, outerName, fieldName, true)
+          case tpe if (tpe.typeSymbol.isClass && tpe.typeSymbol.asClass.isCaseClass) =>
+            val prefix = outerName.map(pre => s"$pre$fieldName.")
+            expandMethod(tpe, prefix, isOption)
+          case tpe if allowUnknownTypes => simpleRet
+          case _ => c.abort(c.enclosingPosition, s"Case class ${T} is not pure primitives or nested case classes")
+        }
+      }
 
-    def expandMethod(outerTpe: Type, outerName: Option[String], isOption: Boolean): List[(Tree, String)] = {
-      outerTpe
-        .declarations
-        .collect { case m: MethodSymbol if m.isCaseAccessor => m }
-        .flatMap { accessorMethod =>
-          val fieldName = accessorMethod.name.toTermName.toString
-          val fieldType = accessorMethod.returnType
-          matchField(fieldType, outerName, fieldName, isOption)
-        }.toList
-    }
+      def expandMethod(outerTpe: Type, outerName: Option[String], isOption: Boolean): List[(Tree, String)] = {
+        outerTpe
+          .declarations
+          .collect { case m: MethodSymbol if m.isCaseAccessor => m }
+          .flatMap { accessorMethod =>
+            val fieldName = accessorMethod.name.toTermName.toString
+            val fieldType = accessorMethod.returnType.asSeenFrom(outerTpe, outerTpe.typeSymbol.asClass)
+
+            matchField(fieldType, outerName, fieldName, isOption)
+          }.toList
+      }
 
-    val prefix = if (namingScheme == NamedNoPrefix) None else Some("")
-    val expanded = expandMethod(T.tpe, prefix, false)
-    if (expanded.isEmpty) c.abort(c.enclosingPosition, s"Case class ${T} has no primitive types we were able to extract")
+      val prefix = if (namingScheme == NamedNoPrefix) None else Some("")
+      val expanded = expandMethod(T.tpe, prefix, false)
+      if (expanded.isEmpty) c.abort(c.enclosingPosition, s"Case class ${T} has no primitive types we were able to extract")
 
-    val typeTrees = expanded.map(_._1)
+      expanded
+    }
 
+    val typeTrees = expanded.map(_._1)
     val res = if (namingScheme == NamedWithPrefix || namingScheme == NamedNoPrefix) {
       val fieldNames = expanded.map(_._2)
```

Comment:
note this is cast to `Comparable` below, so actually using `Comparable[_]` instead of `String` would have worked, I think.